### DISCUSSION TITLE
Plugin affecting SignalRGB effects (or animation) smoothness

### DISCUSSION THREAD

**Discussion by xxxtemp1xxx** [EvanMulawski/FanControl.CorsairLink#68] [Q&A] [answered]
Hi Evan, first of all thank you so much for this plugin!

As per title, I'm encountering an issue where, once I enable this plugin with FanControl to integrate my Corsair devices, it seems to affect the SignalRGB effects or animations by making it less smooth and more choppy/jumpy, if you know what I mean.

For example, I have a default "rainbow" effect in SignalRGB for my Corsair ML (8-LED) fans. With only SignalRGB running, the effects are very smooth — i.e. the rainbow RGB seamlessly blends from one color to the next, just like any default lighting when you first install the fans without any software.

But once I open FanControl with this plugin, the rainbow effect loses that seamless gradation and becomes choppy and jumpy, as if the frame rate has drastically decreased. The colors "jump" from one color to the next — more or less every 1 second or so — giving me reason to believe that the polling between FanControl (and this plugin) and the Corsair devices are somehow disrupting the RGB effects being communicated between SignalRGB and the Corsair devices.

I was hoping that, while both SignalRGB and FanControl may talk to the Corsair devices simultaneously, the former would only talk RGB while the latter would only talk fan control, thereby not disrupting one another. Is this not the case?

So my questions are:

1. Do you think the RGB effects being affected in such manner is indeed because of how this plugin (and FanControl) communicates with the Corsair devices alongside SignalRGB?
2. Is there any workaround for this? Such as increasing the "polling rate" to make the RGB effects more smooth or perhaps not affecting RGB at all? 

I've been using iCUE all this time but would really like to try using both FanControl (with this plugin) and SignalRGB together, without them conflicting/disrupting one another. Or is this more of a hardware issue on Corsair's side where it can't talk to different software simultaneously.

As a side note, iCUE is installed but it's completely disabled and there are no services running in the background when I run FanControl and/or SignalRGB.

**Comment by Fr0stX76** [comment]
This issue was answered in this thread

https://github.com/EvanMulawski/FanControl.CorsairLink/issues/65

Sadly, data is data, regardless it's nature (fan, rgb, monitoring...). And RGB effects send a lot of commands per seconds. The mutex used by the Plugin and SignalRGB prevent overlap of commands from multiple software to the device (which without, would throw errors or worst, freeze), but at the cost of latency since the mutex "lock" one software while the other one talk (in milliseconds...)

When FanContol Plugin is talking, signalRGB is muted for a couple ms, which in the end result in the choppiness of the effect you are experiencing.

Sadly from a software point of view, it is an hardware limitation. As a user, you could try reducing the effect intensity (if SignalRGB has an effect speed slider) which would reduce the amount of update/command sent to the device. Or use a static color... Which is probably a deal breaker for most users. Apart from those limited options I don't think anything else can be done.

**Reply by xxxtemp1xxx** [reply]
Thanks for sharing the link, @DfrostX !

After reading through it, I can confirm the stuttering issue explained in that thread is pretty much identical to mine, although I would say the stuttering or choppy lighting effect occurs every 1 second or so (which I suspect has to do with the polling rate of the monitoring sensors).

Unfortunately, the RGB stuttering remains pretty visible even with a basic Rainbow (classic) effect with the Speed set to 1 (the slowest).

While I'll await for Evan to chip in, this very much looks like an inevitable limitation with the way the Corsair hardware is able to talk with multiple software for RGB and fan control. I wonder if there is a way for the hardware to talk in "separate lines" — one for RGB and one for fan control.

In any matter, as it stands now this is truly a bummer because I usually opt for colorful effects (rather than a simple solid color) and having the colors jump from one to the next every second in my field of view is a bit too distracting to bear on a daily basis. I may have to hold off on using FanControl together with SignalRGB unless there are workarounds to this problem.

**Reply by EvanMulawski** [reply, marked as answer]
Hi @iXzenoS, @DfrostX is correct. Since SignalRGB is constantly sending RGB data to the device, interrupting it for just a few milliseconds will cause a stutter if the animation is fast enough. The only "solution" is to tune the animation such that it isn't as noticeable. It's unfortunate but it's the best we can do.

Also, there is only one "channel" for communication - no separation of lighting and fan control communication - since Corsair intends for everyone to use iCUE.

**Reply by xxxtemp1xxx** [reply]
Thanks, @EvanMulawski for the confirmation. It's unfortunate, but it is what it is I guess.

The good news is I just did some analysis and realized that iCUE isn't as bad as a lot of people think it to be. Without using its "murals" feature which apparently uses a lot of CPU (I haven't tested this), all the other effects and fan controls use a similar amount of resources as the combination of FanControl + SignalRGB.

Looks like I'll be sticking with iCUE for now as it ended up being most simple and gets the job done on my PC. All the best.
